docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1259 @@
+
+### Recommended number of nodes and zones
+
+It is recommended that for high-availability setup an odd number of nodes (node tolerance) or zones (zone tolerance) must be used. This also follows the [recommendations](https://etcd.io/docs/v3.4/faq/#why-an-odd-number-of-cluster-members) on the etcd cluster size. The recommendations for number of zones will be largely influenced by quorum-based etcd cluster setup recommendations as other shoot control plane components are either stateless or non-quorum-based stateful components.
+
+**Let's take the following example to explain this recommendation further:**
+* Seed clusters' worker nodes are spread across **two zones**
+* Gardener would distribute a **three member** etcd cluster - AZ-1: 2 replicas, AZ-2: 1 replica
+* If AZ-1 goes down then, quorum is lost and the only remaining etcd member enters into a read-only state.
+* If AZ-2 goes down then:
+  * If the leader is in AZ-2, then it will force a re-election and the quorum will be restored with 2 etcd members in AZ-1.
+  * If the leader is not in AZ-2, then etcd cluster will still be operational without any downtime as the quorum is not lost.
+
+**Result:**
+* There seems to be no clear benefit to spreading an etcd cluster across 2 zones as there is an additional cost of cross-zonal traffic that will be incurred due to communication amongst the etcd members and also due to API server communication with an etcd member across zones.
+* There is no significant gain in availability as compared to an etcd cluster provisioned within a single zone. Therefore it is a recommendation that for regions having 2 availability zones, etcd cluster should only be spread across nodes in a single AZ.
+
+### Recommended number of replicas
+
+The minimum number of replicas required to achieve HA depends on the [topology](#topologies) and the requirement of each component that run in an `active-active` mode.
+
+**Active-Active**
+* If application needs a quorum to operate (e.g. `etcd`), at least **three replicas** are required ([ref](https://etcd.io/docs/v3.3/faq/#why-an-odd-number-of-cluster-members)).
+* Non-quorum based components are also supposed to run with a minimum count of **three replicas** to survive node/zone outages and support load balancing.
+
+**Active-Passive**
+* Components running in  a `active-passive` mode are expected to have at least **two** replicas, so that there is always one replica on stand-by.
+
+## Gardener Shoot API
+
+### Proposed changes
+
+The following changes to the shoot API are suggested to enable the HA feature for a shoot cluster:
+
+```yaml
+kind: Shoot
+apiVersion: core.gardener.cloud/v1beta1
+spec:
+  controlPlane:
+    highAvailability:
+      enabled: true # by default HA is disabled.
+      failureDomain:
+        type: <single-zone | multi-zone>
+        whenUnsatisfied: <DoNotSchedule | ScheduleAnyway>
```

Comment:
Why would we introduce `controlPlane.highAvailability` and not only `highAvailability`? What's the advantage of the extra level?

I don't like or at least understand `whenUnsatisfied`. Somewhere in the world, a landscape will probably always have at least 1 single-zone seed and 1 multi-zone seed, so when will `DoNotSchedule` really take effect? Besides, I doubt people will specify that and then don't care (`ScheduleAnyway`). From my PoV, I would drop `whenUnsatisfied` and have `failureDomain: <single-zone | multi-zone>`. We select the "nearest" fitting cluster then (filtering, no scoring), i.e. `single-zone` fits everywhere (if the seed has capacity), but `multi-zone` only on multi-zone seeds (single-zone seeds are filtered out). From the remaining set of seeds, we pick the nearest with the least capacity. No?

Reply:
> Why would we introduce `controlPlane.highAvailability` and not only `highAvailability`? What's the advantage of the extra level?

I think `controlPlane.highAvailability` makes it clear that the options HA options (`single-zone`, `multi-zone`) are for the control plane (and not for the Shoot cluster itself).

Reply:
Okay, then let's drop `whenUnsatisfied` 👍 Does it still make sense to have the extra level that gives us the freedom to define further options in the future? (maybe `whenUnsatisfied` becomes a topic again)

Also, what do you think about defining which is actually more precise?
```
failureDomain:
  type: <node | zone>
```

Reply:
Please see the latest version, changes are reflected there.